WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt

```diff
@@ -97,74 +93,61 @@ class UploadStarter @Inject constructor(
     fun activateAutoUploading(processLifecycleOwner: ProcessLifecycleOwner) {
         // We're skipping the first emitted value because the processLifecycleObserver below will also trigger an
         // immediate upload.
-        connectionStatus.skip(1).observe(processLifecycleOwner, Observer {
-            queueUploadFromAllSites()
-        })
+        connectionStatus.skip(1).observe(processLifecycleOwner) { queueUploadFromAllSites() }
 
         processLifecycleOwner.lifecycle.addObserver(processLifecycleObserver)
     }
 
-    fun queueUploadFromAllSites() = launch {
-        val sites = siteStore.sites
-        try {
-            checkConnectionAndUpload(sites = sites)
-        } catch (e: Exception) {
-            AppLog.e(T.MEDIA, e)
-        }
-    }
+    fun queueUploadFromAllSites() = launch { checkConnectionAndUpload(sites = siteStore.sites) }
 
     /**
      * Upload all local drafts from the given [site].
      */
-    fun queueUploadFromSite(site: SiteModel) = launch {
-        try {
-            checkConnectionAndUpload(sites = listOf(site))
-        } catch (e: Exception) {
-            AppLog.e(T.MEDIA, e)
-        }
-    }
+    fun queueUploadFromSite(site: SiteModel) = launch { checkConnectionAndUpload(sites = listOf(site)) }
 
     /**
      * If there is an internet connection, uploads all posts with local changes belonging to [sites].
      *
      * This coroutine will suspend until all the [upload] operations have completed. If one of them fails, all query
-     * and queuing attempts ([upload]) will be canceled. The exception will be thrown by this method.
+     * and queuing attempts ([upload]) will continue. The last exception will be thrown by this method.
      */
     private suspend fun checkConnectionAndUpload(sites: List<SiteModel>) = coroutineScope {
-        if (!networkUtilsWrapper.isNetworkAvailable()) {
-            return@coroutineScope
-        }
-
-        sites.forEach {
-            launch(ioDispatcher) {
-                upload(site = it)
+        if (!networkUtilsWrapper.isNetworkAvailable()) return@coroutineScope
+        try {
+            sites.forEach {
+                launch(ioDispatcher) {
+                    upload(site = it)
+                }
             }
+        } catch (e: Exception) {
```

Comment:
This is a good point, and I think this actually might shed light on the original issue. I see that this same possibility exists in [an earlier version of this implementation](https://github.com/wordpress-mobile/WordPress-Android/blob/749de373db7afee26b2214f2934ba82c377a90e4/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt#L138), and I believe that the underlying "double-unlock" issue could be a result of this.

One theory / possibility (after reading the backscroll, and this comment) is that the `forEach` is spawning multiple jobs on the `ioDispatcher` (with a thread pool), most of them blocking on the `lock()` invocation. Then, if any of the child jobs throws, the parent job cancels, and all siblings too, and then things get a little bit tricky, imo. Since `lock()` docs say:

> **This function releases the lock if it was already acquired by this function before the CancellationException was thrown.

and

> **Note that this function does not check for cancellation when it is not suspended.**

So, although each site's uploads should be blocked and served in FIFO manner, I wonder if the cancellation bubbling up to the parent job and then back down allows another child job to acquire the lock before it is cancelled as a sibling :thinking: 

Though there is a "prompt cancellation guarantee", does this guarantee that it cancels immediately when a sibling cancels? (Since those children can be on different threads, along with the fact that we catch and re-throw, it makes me wonder).

I have not tested this hypothesis, so please take it with a grain of salt. I just wanted to suggest it as I've skimmed the back scroll, and your comment seems that it might shed some light on the issue.

Reply:
Yeah, I was actually trying to reproduce it with some custom code (though I was running it as a Unit test on the JVM and not on an actual device, so I believe thread implementation might be different).

Because of the `mutex`, no upload jobs will be _running_ at the same time, though their coroutines would've been started and they're all suspended at the `lock` function. I think it is _possible_

> I wonder if the cancellation bubbling up to the parent job and then back down allows another child job to acquire the lock before it is cancelled as a sibling 🤔

In my tests, I see this definitely happens, though it doesn't seem to cause of the crash because in that scenario the `unlock` call throwing an exception would be either [this](https://github.com/Kotlin/kotlinx.coroutines/blob/81e17dd37003a7105e542eb725f51ee0dc353354/kotlinx-coroutines-core/common/src/sync/Mutex.kt#L116) or [this](https://github.com/wordpress-mobile/WordPress-Android/blob/1b5e1f3fc97e13d0068ea803d5df7777f421180b/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt#LL186C1-L186C1), but not the one causing the Sentry crash, which is [this one](https://github.com/Kotlin/kotlinx.coroutines/blob/81e17dd37003a7105e542eb725f51ee0dc353354/kotlinx-coroutines-core/common/src/sync/Mutex.kt#L388).

Since the crash happens, I am guessing is also possible that timing can make that be called right when one of the jobs released the lock AND while the cancellation is reaching the [internal continuation](https://github.com/Kotlin/kotlinx.coroutines/blob/81e17dd37003a7105e542eb725f51ee0dc353354/kotlinx-coroutines-core/common/src/sync/Mutex.kt#L189) in `suspendCancellableCoroutineReusable` from the Mutex.

This is a tricky crash, and I still think it _could_ happen even using just the `withLock` and not calling `unlock` on our side, though [that extra](https://github.com/wordpress-mobile/WordPress-Android/blob/1b5e1f3fc97e13d0068ea803d5df7777f421180b/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadStarter.kt#LL186C1-L186C1) `unlock` on our side definitely doesn't help, so maybe @ovitrif changes here help solve the issue.

I feel that updating to coroutines to 1.7.0 will definitely solve this crash (since the Mutex internals were completely changed) though not sure what would take to make that dependency update.

Reply:
> Since the crash happens, I am guessing is also possible that timing can make that be called right when one of the jobs released the lock AND while the cancellation is reaching the [internal continuation](https://github.com/Kotlin/kotlinx.coroutines/blob/81e17dd37003a7105e542eb725f51ee0dc353354/kotlinx-coroutines-core/common/src/sync/Mutex.kt#L189) in suspendCancellableCoroutineReusable from the Mutex.

Yeah, this is exactly what I was thinking might be happening (and thanks for linking to the various sources where `unlock` is called). It is possible that the test dispatcher exhibits different behavior from the `ioDispatcher` actually used in production. If it is indeed a timing issue / race condition of some sort, this might explain why it is hard to reproduce in the test environment.

> This is a tricky crash, and I still think it could happen even using just the `withLock `

I also agree, since, as @ovitrif pointed out, it is basically doing the same thing we were doing (i.e. calling `unlock` in a finally).

> In this scenario it feels to me the mutex (and concurrency) is not needed since in practice we are just running each job sequentially.
> 
> It looks like all those jobs could be called from 1 coroutine, with some exception handling, instead of relying on concurrent jobs and a syncing mechanism (mutex in this case).

I agree with this as well, especially since the jobs are blocked at the highest level, before any control flow traverses into the more granular layers of the implementation where service boundaries might offer concurrency advantages (disk vs. network performance characteristics).

Also, considering performance from a high level, if we hypothetically implemented more granular concurrency, I have doubts that such advantages would be significant anyway, since the likely bottleneck would be the device's network bandwidth (in most circumstances) - so uploading to multiple sites simultaneously likely won't be much (if at all) faster than uploading to each one sequentially, and incurs this complexity cost, as well as an increased risk of multiple interrupted transfers instead of just one in the case of network failure.

That said, currently, it seems that the purpose of the `mutex` is described in the code comment:

```kotlin
    /**
     * When the app comes to foreground both `queueUploadFromAllSites` and `queueUploadFromSite` are invoked.
     * The problem is that they can run in parallel and `uploadServiceFacade.isPostUploadingOrQueued(it)` might return
     * out-of-date result and a same post is added twice.
     */
```

so maybe more work needs to be done to make those `queueUploadFrom{AllSites,Site}` methods idempotent? I don't have enough context to know what problem this `mutex` solves here. Perhaps the "deduping" should be happening at the layer closer to the queue, with any needed synchronization there? @ParaskP7 :wave: :smile:  I see that this code comment was from https://github.com/wordpress-mobile/WordPress-Android/commit/43949f11e88fe29b3710a7812f6c1e03fc18f2fc but it appears originate a bit earlier: (https://github.com/wordpress-mobile/WordPress-Android/pull/10878) - so I'm not sure if you have full context of the original reason for that comment. But I wonder wdyt about this? I.e. do you think it is possible this synchronization / deduping could be done in the service somehow?

Reply:
Thanks a lot for the input @thomashorta and @mkevins 🙇🏻 , you're awesome and really helpful 🙏🏻 

> so maybe more work needs to be done to make those queueUploadFrom{AllSites,Site} methods idempotent? I don't have enough context to know what problem this mutex solves here

If I can think outside the box on this one, this is also my view. That `mutex` could be replaced with a different solution that avoids the post duplication 👍🏻